ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+
+  let set_vlower ?log v vlower =
+    (match log with
+    | None -> ()
+    | Some log -> log := Cvlower (v, v.vlower) :: !log);
+    v.vlower <- vlower
+
+  let submode_cv : type a. ?log:_ -> a C.obj -> a -> a var -> (unit, a) Result.t
+      =
+    fun (type a) ?log (obj : a C.obj) a' v ->
+     if C.le obj a' v.lower
+     then Ok ()
+     else if not (C.le obj a' v.upper)
+     then Error v.upper
+     else (
+       update_lower ?log obj v a';
+       if C.le obj v.upper v.lower then set_vlower ?log v [];
+       Ok ())
+
+  let submode_cmv :
+      type a l.
+      ?log:_ -> a C.obj -> a -> (a, l * allowed) morphvar -> (unit, a) Result.t
+      =
+   fun ?log obj a (Amorphvar (v, f) as mv) ->
+    (* Want a <= f v, therefore f' a <= v. Ideally the two are equivalent.
+       However, [f v] could have been implicitly injected into a larger lattice,
+       which means [a] could be outside of [f]'s co-domain, which is also [f']'s
+       domain - so applying [f'] to [a] could be invalid.
+
+       The following (seemingly redundant) several lines prevents that:
+       - If a <= (f v).lower, immediately succeed
+       - If not (a <= (f v).upper), immediately fail
+       - Note that at this point, we still can't ensure that a >= (f v).lower.
+         (We don't assume total ordering for generality.)
+        Therefore, we set a = join a (f v).lower. This operation has no effect
+        in terms of submoding, but we have now ensured that a is within the
+        range of f v, and thus a is within the co-domain of f, and thus its
+        safe to apply f' to a. *)
+    if C.le obj a (mlower obj mv)
+    then Ok ()
+    else if not (C.le obj a (mupper obj mv))
+    then Error (mupper obj mv)
+    else
+      let a = C.join obj a (mlower obj mv) in
+      let f' = C.left_adjoint obj f in
+      let src = C.src obj f in
+      let a' = C.apply src f' a in
+      assert (Result.is_ok (submode_cv ?log src a' v));
+      Ok ()
+
+  (* In the following, functions might take two logs: [log] records changes
+     caused by the operation, [optim] records optimizing changes not caused by the
+     operation. Semantically, [optim] happens before [log] and doesn't rely on [log].
+     On the other hand, some changes in [log] might rely on [optim]. *)
+
+  (** Returns [None] if success; [Some x] if failed, and [x] is the next best
+     guess to replace the constant argument that MIGHT succeed. *)
+  let rec submode_vc :
+      type a. ?log:_ -> a C.obj -> a var -> a -> (unit, a) Result.t =
+    fun (type a) ?log (obj : a C.obj) v a' ->
+     if C.le obj v.upper a'
+     then Ok ()
+     else if not (C.le obj v.lower a')
+     then Error v.lower
+     else (
+       update_upper ?log obj v a';
+       let r =
+         v.vlower
+         |> find_error (fun mu ->
+                let r = submode_mvc ?log obj mu a' in
+                (if Result.is_ok r
+                then
+                  (* Optimization: update [v.lower] based on [mlower u].*)
```

Comment:
I think we do it here, because we are iterating the `vlower` and recursively calling `submode_mvc`  anyway (essentially traversing the subtree). So this optimization shouldn't add much time to that. 

The morphism chain in `mlower` should not be very long - at most six in our experience.